### PR TITLE
Porting to GHEX 0.2.0 + fixes for multiple devices

### PR DESCRIPTION
**Pull request by bettiolm**
- `ghex_comm` runtime ported to ghex 0.2.0, using oomph as transport library.
- fixed an issue that caused `detected error differences between the runs` when using the `ghex_comm` runtime on multiple devices. It was due both to streams synchronization issues and to the choice of the default set-device strategy.
- GHEX code does not depend anymore on the `gridtools::_impl` namespace.
- NOTE: semantics of the `--device-mapping` option has changed: it now accepts only one device id per rank, meaning that different threads on the same rank must use the same device.
- AMD GPUs fixes are not included, and will follow shortly.

### CODE REVIEW

**Comment by bettiolm** [comment]
AMD GPUs support for `ghex_comm` runtime has been added. The only feature that cannot be fully transferred to AMD GPUs is the `--default-stream=per-thread` compile option. It seems the feature will be supported in future versions of the clang compiler (see e.g. https://github.com/llvm/llvm-project/blob/llvmorg-16-init/clang/include/clang/Driver/Options.td#L981 , from which it seems it will be included in `clang-16`).

**Review by fthaler — CHANGES_REQUESTED** [review]
Thank you very much for this work! I added some comments about the `// TODOs` still in the PR.

File: src/runtime/ghex_comm/run.cpp [hunk 1]
```diff
@@ -295,29 +283,28 @@ class grid::impl {
...
-      using arch_t = gt::ghex::gpu;
+      using arch_t = ghex::gpu;
 #else
-      using arch_t = gt::ghex::cpu;
+      using arch_t = ghex::cpu;
 #endif
-      auto field =
-          gt::ghex::wrap_gt_field<arch_t>(domain, storage, {halo, halo, 0});
+      auto field = ghex::wrap_gt_field<arch_t>(
+          domain, storage, {halo, halo, 0}); // TO DO: device id
```
[the remark this answers is not in the document]
Comment (bettiolm) [reply src/runtime/ghex_comm/run.cpp]: Just a leftover. It has been fixed on the GHEX side. We don't pass the device id here, and ghex uses the current device id as default value. I updated the comment.

File: src/runtime/ghex_comm/run.cpp [hunk 2]
```diff
@@ -212,9 +201,9 @@ class grid::impl {
   coordinate_type m_first;
   coordinate_type m_last;
   domain_vec m_domains;
-  context_ptr_t m_context;
+  context_ptr_t m_context; // TO DO: is a pointer still needed here?
```
[the remark this answers is not in the document]
Comment (bettiolm) [reply src/runtime/ghex_comm/run.cpp]: Sure, the context is needed. Probably I was not sure whether to store the pointer here. I removed the comment.

File: src/runtime/ghex_comm/run.cpp [hunk 3]
```diff
@@ -17,14 +17,7 @@
...
 
-#ifdef GTBENCH_USE_GHEX_UCP
-#include <ghex/transport_layer/ucx/context.hpp>
-using transport = gridtools::ghex::tl::ucx_tag;
-#else
-#include <ghex/transport_layer/mpi/context.hpp>
-using transport = gridtools::ghex::tl::mpi_tag;
-#endif
-#include <ghex/transport_layer/util/barrier.hpp>
+// TO DO: barrier
```
[the remark this answers is not in the document]
Comment (bettiolm) [reply src/runtime/ghex_comm/run.cpp:20]: Shouldn't be needed anymore with the new implementation. The benchmark is working even without. Let me run one last test to make sure.

**Comment by fthaler** [comment]
Looks good to me as soon as the barrier question is resolved and respective comments are cleaned. But as I do not have much knowledge of GHEX, maybe @boeschf can also do a brief review?

**Review by boeschf — APPROVED** [review]

File: src/runtime/ghex_comm/run.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (bettiolm) [reply src/runtime/ghex_comm/run.cpp]: I eventually restored the barrier.

File: src/runtime/ghex_comm/run.cpp [hunk 4]
```diff
@@ -212,9 +201,9 @@ class grid::impl {
   coordinate_type m_first;
   coordinate_type m_last;
   domain_vec m_domains;
-  context_ptr_t m_context;
+  context_ptr_t m_context; // TO DO: is a pointer still needed here?
   patterns_ptr_t m_patterns;
-  gridtools::ghex::tl::barrier_t m_barrier;
+  // TO DO: m_barrier;
```
[the remark this answers is not in the document]
Comment (bettiolm) [reply src/runtime/ghex_comm/run.cpp]: Done

File: src/runtime/ghex_comm/run.cpp [hunk 5]
```diff
@@ -224,7 +213,9 @@ class grid::impl {
                                 (int)global_resolution.y - 1,
                                 (int)global_resolution.z - 1}},
         m_global_resolution{global_resolution.x, global_resolution.y},
-        m_barrier(num_sub_domains) {
+        m_context{std::make_unique<context_t>(MPI_COMM_WORLD,
+                                              (num_sub_domains > 1))} {
+    // TO DO: m_barrier(num_sub_domains) {
```
[the remark this answers is not in the document]
Comment (bettiolm) [reply src/runtime/ghex_comm/run.cpp]: Done

File: src/runtime/ghex_comm/run.cpp [hunk 6]
```diff
@@ -271,20 +262,17 @@ class grid::impl {
...
   }
 
   impl(impl const &) = delete;
   impl &operator=(impl const &) = delete;
 
   sub_grid operator[](unsigned int i) {
     const auto &dom = m_domains[i];
-    auto comm = m_context->get_communicator();
-    m_barrier(comm);
+    // TO DO: m_barrier(comm);
```
[the remark this answers is not in the document]
Comment (bettiolm) [reply src/runtime/ghex_comm/run.cpp]: Done

**Comment by bettiolm** [comment]
@fthaler, I addressed your comments and reintroduced the barrier, as in the original implementation.

**Review by fthaler — APPROVED** [review]

**Comment by fthaler** [comment]
@bettiolm Thank you very much!

**Pull request merged**